osu.Game/Online/Multiplayer/MultiplayerClient.cs

```diff
@@ -50,80 +54,66 @@ private void apiStateChanged(ValueChangedEvent<APIState> state)
             {
                 case APIState.Failing:
                 case APIState.Offline:
-                    connection?.StopAsync();
-                    connection = null;
+                    Task.Run(() => disconnect(true)).CatchUnobservedExceptions();
                     break;
 
                 case APIState.Online:
-                    Task.Run(Connect);
+                    Task.Run(connect).CatchUnobservedExceptions();
                     break;
             }
         }
 
-        protected virtual async Task Connect()
+        private async Task connect()
         {
-            if (connection != null)
-                return;
+            cancelExistingConnect();
 
-            connection = new HubConnectionBuilder()
-                         .WithUrl(endpoint, options =>
-                         {
-                             options.Headers.Add("Authorization", $"Bearer {api.AccessToken}");
-                         })
-                         .AddNewtonsoftJsonProtocol(options => { options.PayloadSerializerSettings.ReferenceLoopHandling = ReferenceLoopHandling.Ignore; })
-                         .Build();
+            await connectionLock.WaitAsync(10000);
```

Comment:
Is this intended to silently continue if it fails to acquire? Perhaps you want to make use of the bool return?

Reply:
ah, mistook this to throw on failure. good catch.